xla/hlo/transforms/collectives/rematerialize_large_all_gather.cc

```diff
@@ -0,0 +1,165 @@
+/* Copyright 2024 The OpenXLA Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "xla/hlo/transforms/collectives/rematerialize_large_all_gather.h"
+
+#include "xla/hlo/ir/hlo_casting_utils.h"
+#include "xla/hlo/ir/hlo_instruction.h"
+#include "xla/hlo/ir/hlo_instructions.h"
+#include "xla/hlo/ir/hlo_opcode.h"
+#include "xla/hlo/utils/hlo_query.h"
+#include "xla/hlo/transforms/simplifiers/hlo_dce.h"
+#include "xla/service/pattern_matcher.h"
+#include "xla/shape_util.h"
+
+namespace xla {
+
+std::pair<bool, std::vector<ReplicaGroup>>
+RematerializeLargeAllGather::GetTpReplicaGroup(HloComputation* computation) {
```

Comment:
This looks like a rather specific pattern to me. Is there a meaningful way to express the targeted pattern more generally?

Reply:
Hmmm, that is the sequence parallel pattern.

I am not aware of a pattern besides this one that can differentiate between sequence parallel all-gathers and data parallel all-gathers. We only want sequence-parallel here.